In [1]:
!pip install nltk

In [2]:
import torch
import torch.nn as nn
import torch.optim as optim
import numpy as np
from collections import Counter
import nltk
from nltk.tokenize import word_tokenize
import random
from torch.utils.data import Dataset, DataLoader

In [3]:
document = """
Natural language processing (NLP) is a subfield of artificial intelligence (AI).
It focuses on enabling computers to understand, interpret, and generate human language.
NLP combines computer science, linguistics, and machine learning.
It's used in various applications like machine translation, sentiment analysis, and chatbots.
My friend like books every day
Our teacher read cricket during holidays
I play cricket after school
My friend visit movies after school
The cat play the beach at night
We visit bananas at night
We watch football when it rains
The cat read the beach with friends
He read cricket every day
My friend eat apples before dinner
She hate books before dinner
My friend hate movies before dinner
I hate apples every day
The cat enjoy football at night
The dog eat bananas on weekends
He visit the beach on weekends
The dog like books with friends
We watch apples with friends
The cat like books before dinner
We eat the park every day
My friend watch the beach during holidays
Our teacher love cricket during holidays
You watch bananas after school
I enjoy apples on weekends
My friend read movies in the morning
Our teacher prefer apples every day
The dog play the market every day
The cat watch books every day
They love the park in the summer
The cat eat books in the summer
The dog love movies every day
My friend read books at night
She like the beach every day
I eat apples during holidays
You watch the beach with friends
He eat movies before dinner
My friend read music before dinner
He hate movies with friends
We eat movies during holidays
Our teacher watch the market every day
He hate movies on weekends
The cat visit apples before dinner
I watch football on weekends
The cat read the park with friends
My friend read books on weekends
Our teacher play apples when it rains
He watch the beach on weekends
Our teacher enjoy books with friends
My friend visit books at night
The dog watch the beach when it rains
I watch music when it rains
They love music every day
They prefer apples when it rains
He prefer the market every day
The cat love apples when it rains
She enjoy the beach in the summer
The cat play apples after school
My friend read cricket every day
He read the beach with friends
He like the market on weekends
They eat movies in the summer
Our teacher love the park on weekends
She watch the park at night
The dog love cricket with friends
Our teacher visit music in the morning
Our teacher love cricket when it rains
She like the park during holidays
We enjoy movies on weekends
Our teacher hate apples at night
The dog hate apples when it rains
I watch football in the summer
The dog hate football at night
You eat cricket when it rains
They play music with friends
My friend eat music at night
I like music when it rains
He play movies every day
You hate movies during holidays
We watch movies after school
The dog hate movies before dinner
He enjoy the beach during holidays
I play the park at night
I watch football during holidays
We prefer books when it rains
He prefer the market with friends
The cat prefer apples when it rains
My friend love the beach when it rains
The dog eat apples with friends
They play cricket when it rains
Our teacher enjoy bananas in the morning
Our teacher visit movies when it rains
Our teacher visit the market with friends
You eat books in the morning
They eat bananas in the summer
The dog read cricket on weekends
The dog enjoy the market when it rains
She like the park every day
My friend watch music in the summer
Our teacher love football at night
My friend play bananas during holidays
You watch bananas when it rains
I hate apples every day
Our teacher prefer the market at night
We like the market in the summer
They love music before dinner
I visit cricket with friends
I hate cricket with friends
We eat books in the summer
She eat apples at night
Our teacher watch bananas before dinner
You eat football on weekends
The cat play movies in the morning
We eat the park every day
She love books on weekends
They hate football at night
We like cricket before dinner
Our teacher read the park on weekends
He like bananas during holidays
I eat the beach in the summer
I love movies in the morning
My friend read cricket in the summer
He eat bananas with friends
You read the beach before dinner
She eat music every day
The dog like the market in the morning
He watch apples in the summer
The dog visit bananas at night
She enjoy cricket on weekends
He enjoy the beach on weekends
My friend love cricket before dinner
You read the beach after school
We prefer the market when it rains
She like cricket in the morning
I watch the market after school
She play the park after school
Our teacher hate bananas with friends
I love music every day
They play football in the morning
I watch movies after school
I prefer the park on weekends
My friend eat the market with friends
You visit movies after school
The cat enjoy football after school
The dog hate apples in the morning
My friend like the beach on weekends
He play books at night
She love bananas in the summer
She like apples before dinner
We play music with friends
She play bananas on weekends
My friend read apples after school
The dog eat the market in the summer
My friend watch the park before dinner
He watch bananas in the summer
The dog watch apples every day
She play books in the summer
The dog visit the park after school
I read bananas during holidays
She prefer bananas every day
He eat the market after school
They eat the park at night
I love football at night
He watch the market before dinner
My friend like football on weekends
You watch apples after school
We watch football before dinner
My friend enjoy apples at night
Our teacher watch the market before dinner
Our teacher visit music after school
Our teacher enjoy football after school
You play football with friends
He play the market in the morning
You hate bananas when it rains
You eat football before dinner
We like books in the summer
I visit apples with friends
He watch football with friends
I read the beach at night
He eat the park during holidays
You enjoy the market with friends
He enjoy the beach when it rains
The dog watch music with friends
The dog hate apples in the summer
We read music after school
She visit books with friends
She read books during holidays
I prefer movies after school
Our teacher read football with friends
He prefer the park every day
She hate books with friends
The dog like movies in the morning
We hate the market in the morning
We play bananas before dinner
The dog like books with friends
My friend prefer the beach when it rains
He watch cricket at night
The cat visit bananas in the morning
They read movies when it rains
My friend enjoy movies after school
The cat eat the market every day
You hate bananas on weekends
We enjoy music before dinner
The cat enjoy the market on weekends
He eat music during holidays
The cat hate bananas when it rains
He enjoy apples after school
Our teacher read the park on weekends
My friend prefer apples in the summer
Our teacher love apples at night
We prefer apples before dinner
We hate apples at night
The dog read apples in the morning
The dog prefer movies in the summer
The dog read football in the summer
They visit the park with friends
You prefer the market in the summer
She watch bananas in the morning
I eat the beach every day
My friend hate the park after school
I read movies before dinner
The dog like the park in the morning
He visit the beach in the morning
He play the beach on weekends
Our teacher read movies every day
We visit apples during holidays
He eat books in the morning
She visit movies during holidays
I eat the park with friends
The cat enjoy the market when it rains
I read movies during holidays
Our teacher read music after school
We love bananas when it rains
My friend eat the park with friends
The dog read books after school
She eat football before dinner
My friend like books during holidays
We watch cricket on weekends
She eat the market when it rains
My friend read football at night
The dog love apples with friends
My friend play the park every day
We hate movies on weekends
Our teacher love books when it rains
He love cricket at night
Our teacher love books during holidays
My friend prefer movies in the summer
The dog eat football in the summer
They like football during holidays
They like the park at night
She play books in the morning
You watch bananas on weekends
She read music before dinner
He enjoy books during holidays
We prefer football with friends
He love bananas in the summer
The dog enjoy bananas when it rains
He read apples after school
Our teacher like the market in the morning
I enjoy apples on weekends
Our teacher visit apples with friends
He like music with friends
Our teacher read the beach at night
My friend eat apples when it rains
Our teacher like the market when it rains
The dog read cricket before dinner
Our teacher like apples after school
She play apples before dinner
Our teacher watch bananas after school
She watch books after school
They eat cricket on weekends
She hate cricket every day
We prefer football with friends
He watch movies before dinner
The dog enjoy bananas during holidays
Our teacher watch the beach after school
She enjoy books after school
You prefer books in the summer
The dog watch cricket with friends
Our teacher visit cricket with friends
You prefer football before dinner
Our teacher watch bananas when it rains
My friend prefer apples when it rains
My friend visit books when it rains
She enjoy bananas when it rains
She hate the beach after school
You play cricket every day
She like music in the morning
You play the beach after school
He read football on weekends
They read books after school
The cat watch the beach after school
The cat hate cricket in the summer
Our teacher play cricket during holidays
The dog visit music in the summer
They visit the market during holidays
You like the park after school
The dog visit books when it rains
Our teacher play the market with friends
My friend visit apples at night
She watch the beach in the morning
We watch bananas every day
The dog enjoy music in the summer
They read books at night
The cat hate apples before dinner
We enjoy football on weekends
Our teacher hate the beach when it rains
My friend play bananas on weekends
My friend like the market on weekends
My friend play music when it rains
She prefer the park during holidays
Our teacher love books before dinner
She visit the park in the summer
He eat movies after school
The dog enjoy music every day
We hate apples in the morning
The cat watch the market at night
He play movies before dinner
He enjoy cricket before dinner
I eat cricket when it rains
They read music with friends
The cat love apples before dinner
You eat apples at night
They play music when it rains
You watch the market at night
Our teacher read apples on weekends
The cat prefer books at night
The dog play apples after school
We love music before dinner
My friend play music in the summer
Our teacher like the market after school
You hate the park at night
We hate football with friends
They eat cricket before dinner
Our teacher enjoy books on weekends
You prefer books when it rains
I like books on weekends
We enjoy the park on weekends
My friend read books in the morning
My friend watch the market when it rains
My friend enjoy cricket with friends
He play movies with friends
My friend prefer cricket on weekends
We read the market when it rains
The dog prefer movies before dinner
They like the park on weekends
I play bananas at night
The dog love cricket on weekends
She visit movies at night
The dog read music with friends
The dog like apples during holidays
She enjoy the park in the summer
She love apples in the morning
The cat prefer football before dinner
He read movies when it rains
You watch the beach during holidays
My friend play the beach before dinner
The dog like movies when it rains
They like books with friends
He love music when it rains
I watch cricket every day
The dog visit books in the summer
Our teacher watch cricket with friends
I hate apples at night
He love the beach every day
My friend prefer the park at night
I play the market during holidays
He love the market with friends
They prefer cricket during holidays
She eat books with friends
The dog visit the park when it rains
I hate apples during holidays
He eat the market with friends
They hate the beach before dinner
The dog eat the park during holidays
He hate the park in the morning
He love the park every day
She read apples when it rains
You enjoy apples with friends
My friend visit the market in the morning
I read books when it rains
My friend watch movies on weekends
We play bananas in the morning
We play music after school
I play books after school
We like the market after school
He hate books in the morning
The dog read music in the morning
I play the market on weekends
Our teacher play bananas on weekends
We hate movies during holidays
They love books in the summer
The cat hate the park after school
The dog eat the park every day
Our teacher enjoy cricket in the summer
She watch the beach during holidays
The cat prefer football on weekends
You visit the beach every day
The dog watch books when it rains
You eat apples in the summer
I read football at night
He love apples on weekends
Our teacher visit the market during holidays
They eat the market after school
Our teacher prefer apples at night
We like apples with friends
You watch football before dinner
You love movies before dinner
The cat watch movies in the summer
The cat enjoy music on weekends
The dog hate apples on weekends
Our teacher watch the beach in the morning
They eat the beach in the morning
He read football with friends
He visit bananas on weekends
The dog watch the market during holidays
Our teacher eat apples on weekends
We watch apples in the morning
We like the market on weekends
They play books on weekends
The cat visit the market with friends
They watch bananas when it rains
I eat apples before dinner
We like music in the summer
Our teacher like the market when it rains
She enjoy apples in the morning
My friend eat the park after school
He visit the beach every day
The dog love football after school
Our teacher visit the market every day
The dog eat the park in the summer
He hate bananas after school
They love books on weekends
We like books in the summer
We eat football on weekends
You like the beach after school
They watch movies with friends
Our teacher watch the market in the morning
The dog like books during holidays
We eat football before dinner
They like the market with friends
Our teacher read the market when it rains
The cat love football before dinner
He like music in the morning
She watch football with friends
She play music during holidays
The cat hate the park during holidays
Our teacher read books before dinner
I play the beach with friends
They play music at night
He love cricket in the morning
She play music when it rains
You enjoy books in the summer
Our teacher visit bananas during holidays
She prefer bananas when it rains
I like movies before dinner
I read the park when it rains
I love cricket at night
The dog visit the market after school
I read cricket every day
I read football at night
My friend visit the park in the summer
The cat eat bananas in the summer
You love the park after school
Our teacher watch football with friends
He visit the beach on weekends
My friend eat books in the morning
Our teacher eat movies in the summer
Our teacher watch apples during holidays
She enjoy the park on weekends
You watch music on weekends
We play music after school
She play the beach during holidays
The cat eat cricket when it rains
The dog watch the market before dinner
The cat eat football during holidays
He prefer music every day
He love the market every day
She eat movies at night
The cat watch the park every day
My friend visit music in the summer
She enjoy music on weekends
You love bananas before dinner
The cat watch the market when it rains
The dog visit music after school
The dog eat the park in the morning
They like music before dinner
You visit movies after school
My friend like the market every day
The cat visit music during holidays
The dog prefer apples with friends
They visit music at night
You enjoy apples after school
They hate music with friends
We hate the park during holidays
We visit the market in the morning
My friend hate the market in the summer
Our teacher like cricket before dinner
We hate movies in the morning
We eat movies with friends
They prefer cricket every day
My friend enjoy movies at night
I watch music with friends
The dog hate books with friends
The cat prefer cricket on weekends
You enjoy movies at night
Our teacher play bananas before dinner
Our teacher watch apples every day
They read the beach after school
You love music before dinner
Our teacher watch the park at night
My friend eat the market with friends
We hate the park every day
She hate the park after school
My friend read the market on weekends
I prefer the beach in the summer
Our teacher eat music after school
I visit the beach in the morning
The dog eat the market every day
I love books in the morning
She visit movies with friends
You play bananas in the summer
We love apples in the summer
They like the park on weekends
He eat the park every day
The cat play cricket in the summer
The cat visit books with friends
You hate the beach before dinner
You enjoy bananas before dinner
You play the park after school
The cat hate movies on weekends
The cat hate the beach with friends
My friend eat bananas with friends
He enjoy the market when it rains
We watch music during holidays
The cat watch the park on weekends
You eat the beach at night
You play music during holidays
They watch cricket at night
She watch the market in the morning
You eat the beach in the summer
My friend visit books in the summer
I eat the market when it rains
The cat enjoy apples every day
The dog read apples on weekends
We play cricket in the summer
The cat eat movies every day
My friend hate football after school
My friend like the market every day
We play cricket during holidays
She read music after school
They like bananas in the summer
They eat the market before dinner
My friend visit music before dinner
The dog enjoy cricket when it rains
I play music before dinner
She read music with friends
You read movies during holidays
We hate cricket every day
My friend love the beach on weekends
I visit the beach with friends
You love apples at night
Our teacher like apples before dinner
We enjoy movies when it rains
I like bananas every day
I read football when it rains
We prefer the beach after school
Our teacher watch movies in the morning
The dog visit the beach on weekends
You like the park when it rains
You play apples every day
We watch football on weekends
We read the market in the morning
They hate movies at night
We read the beach every day
The cat love the market in the morning
My friend eat movies every day
We watch music with friends
The cat read apples when it rains
The cat enjoy football on weekends
I watch the beach when it rains
We play movies before dinner
He eat apples before dinner
The cat enjoy cricket in the summer
You love music in the summer
I watch football with friends. She reads books with friends
The cat play bananas when it rains
I like music after school. He prefers cricket at night
I play movies when it rains
The dog like movies when it rains
The cat hate movies on weekends
I read apples at night
I like books with friends
You enjoy books when it rains
They visit the beach every day
You enjoy cricket in the summer
The cat enjoy music on weekends
My friend visit bananas every day
The dog like apples on weekends
You read the market every day
My friend love football in the summer
The dog enjoy bananas after school
The cat visit books after school
The cat love cricket every day
They enjoy bananas before dinner
The dog prefer movies before dinner
She play movies on weekends
Our teacher eat the beach at night
I read the park with friends
The dog watch the park in the morning
He like books on weekends
You visit the market on weekends
The dog like football in the summer
Our teacher prefer the beach on weekends
I eat the park before dinner
The cat enjoy the market in the summer
My friend read movies in the summer
You love movies before dinner
My friend love cricket in the morning
She play books during holidays
They enjoy cricket with friends
I prefer music during holidays
Our teacher prefer the park in the morning
I watch music on weekends
He enjoy movies in the summer
I play cricket on weekends
I play the market during holidays
The dog love bananas during holidays
Our teacher love bananas when it rains
The cat like the beach after school
They read books on weekends
The cat enjoy the park during holidays
Our teacher visit movies on weekends
The dog read the park in the summer
Our teacher read apples in the morning
The cat enjoy football when it rains
My friend eat cricket during holidays
She prefer the beach before dinner
We visit bananas on weekends
She hate football in the morning
They read movies at night
He like the market in the morning
You play music in the summer
She love movies during holidays
Our teacher love music on weekends
You visit movies in the summer
Our teacher visit movies with friends
I love football with friends
You love apples during holidays
My friend love books before dinner
Our teacher read the beach during holidays
She play books before dinner
He love football with friends
The dog play bananas before dinner
The dog enjoy the park before dinner
They love cricket after school
We read books when it rains
They visit music at night
My friend read football at night
Our teacher like bananas on weekends
My friend read bananas every day
The cat prefer the park on weekends
I like music every day
The dog watch books in the summer
You love music on weekends
My friend eat books when it rains
We visit the beach in the morning
The cat play music before dinner
She visit the beach at night
He eat bananas during holidays
She visit books on weekends
My friend eat apples in the morning
My friend watch football before dinner
She hate football at night
My friend eat bananas after school
We eat books after school
The dog prefer the park when it rains
You prefer football every day
He watch music after school
They like cricket on weekends
We visit music in the summer
She play apples before dinner
My friend prefer movies after school
You hate apples in the morning
He visit football on weekends
You prefer apples in the summer
The cat watch the beach on weekends
My friend love books at night
I visit books with friends
My friend hate cricket when it rains
She prefer the park after school
My friend eat football when it rains
Our teacher enjoy music when it rains
I prefer movies every day
I watch bananas with friends
He watch football every day
The cat prefer music when it rains
She eat music in the summer
Our teacher enjoy apples every day
She love the park in the summer
We watch cricket in the summer
Our teacher watch music before dinner
She visit music in the morning
She like bananas after school
We like cricket in the summer
He watch music every day
I prefer apples on weekends
The dog visit the market in the morning
She prefer cricket when it rains
He read cricket when it rains
My friend watch football when it rains
She enjoy books at night
I eat the market at night
I enjoy music during holidays
My friend play music after school
My friend love football when it rains
He watch the market after school
She play bananas on weekends
I enjoy apples on weekends
She eat the beach after school
She enjoy music before dinner
My friend prefer music before dinner
She hate the park when it rains
I enjoy movies during holidays
Our teacher like apples every day
She visit music when it rains
We play the beach in the summer
We visit the park every day
The cat eat movies after school
She eat the beach on weekends
She prefer bananas during holidays
The dog visit music in the morning
We love cricket with friends
You enjoy bananas with friends
We enjoy the park in the summer
I enjoy apples in the morning
We hate football before dinner
You read the market in the morning
Our teacher visit books in the morning
The dog eat bananas in the summer
You hate movies every day
She love bananas during holidays
She love the park in the morning
Our teacher love the market before dinner
My friend play the beach on weekends
She prefer cricket on weekends
You prefer apples with friends
I hate cricket every day
I play apples with friends
They prefer football in the morning
The cat read football when it rains
She eat cricket in the summer
The dog watch books before dinner
He like music at night
The dog like bananas on weekends
We play apples in the morning
Our teacher enjoy the beach at night
Our teacher love apples when it rains
She play the market every day
I love cricket when it rains
He watch apples in the summer
She watch music when it rains
The cat like books before dinner
You like apples during holidays
We read the market every day
I watch books every day
She visit the market in the summer
I love football when it rains
The dog love apples before dinner
He visit football in the summer
He hate bananas every day
We visit cricket when it rains
The dog eat the beach when it rains
Our teacher like the beach in the summer
She like the park in the morning
Our teacher love apples every day
The dog visit the park with friends
I read the market at night
My friend read music every day
Our teacher enjoy books on weekends
The cat visit the market at night
She play the market on weekends
Our teacher eat bananas every day
The cat enjoy the beach at night
The dog play cricket when it rains
You love cricket with friends
My friend hate football when it rains
The cat eat movies on weekends
My friend eat books every day
We hate movies with friends
He like the market after school
The dog enjoy the market when it rains
They prefer football when it rains
You enjoy football after school
The cat hate cricket on weekends
Our teacher love the market with friends
They visit football in the morning
They read the beach when it rains
They play apples in the summer
The dog love books at night
The dog hate movies when it rains
She enjoy apples with friends
My friend like the market on weekends
You prefer cricket in the summer
He enjoy football in the morning
We visit the market in the morning
The dog read football in the summer
Our teacher like books during holidays
Our teacher love the park with friends
The dog eat the beach when it rains
We love music in the summer
The dog like the beach during holidays
The dog enjoy football during holidays
The dog play cricket before dinner
She love cricket on weekends
He hate the market during holidays
The cat like football when it rains
He like the market when it rains
The dog enjoy bananas at night
We visit the market during holidays
My friend read bananas on weekends
My friend prefer the park during holidays
She read bananas in the summer
The cat prefer movies at night
The dog enjoy movies when it rains
She love bananas on weekends
I like bananas at night
You enjoy movies in the morning
The dog hate apples in the summer
The dog eat music during holidays
You play the market at night
My friend like movies before dinner
I visit the market after school
They prefer the market at night
I play the park at night
The cat eat the park every day
The dog read books at night
The dog watch the market at night
I prefer the market every day
He prefer movies after school
He hate football in the summer
The cat enjoy the beach in the summer
We love cricket in the summer
The cat visit football in the summer
You hate the beach in the morning
The dog visit music after school
You enjoy the market on weekends
She watch cricket after school
I play movies after school
My friend read football before dinner
Our teacher enjoy apples in the morning
He eat the park in the summer
We hate the park on weekends
They eat bananas after school
We love bananas before dinner
They hate music with friends
She love books every day
They prefer the beach at night
They read football every day
I like cricket when it rains
We eat the beach during holidays
You prefer the beach when it rains
He love bananas on weekends
I read the park during holidays
We like cricket in the summer
She hate music every day
I love cricket in the morning
My friend prefer the park when it rains
My friend play apples in the morning
They play cricket in the summer
They play the beach during holidays
I watch cricket before dinner
The cat love the beach at night
He visit cricket in the morning
You prefer cricket with friends
You visit movies when it rains
My friend eat cricket in the summer
We like bananas after school
The cat play football every day
They watch football when it rains
We visit music when it rains
She watch the market in the morning
We hate books in the morning
Our teacher play bananas in the summer
She watch bananas at night
They prefer apples at night
They love the beach at night
We hate the market with friends
The dog visit the market on weekends
I visit the market on weekends
They visit apples every day
I play music during holidays
My friend like music at night
Our teacher hate the market every day
You eat movies every day
My friend eat the park before dinner
He read the beach when it rains
You watch apples when it rains
Our teacher enjoy the beach after school
They eat bananas during holidays
Our teacher eat bananas before dinner
They like bananas before dinner
The cat love the beach after school
Our teacher prefer the market on weekends
She eat bananas every day
They enjoy music during holidays
We love the beach in the morning
My friend prefer books in the summer
I play books in the summer
My friend like the beach in the morning
They watch football every day
I read apples in the summer
Our teacher love movies when it rains
The cat enjoy football after school
They watch books in the summer
The cat hate the market at night
She enjoy cricket during holidays
He like music during holidays
She eat bananas when it rains
You visit bananas in the summer
They watch bananas in the morning
They enjoy cricket before dinner
He visit apples at night
I like apples at night
They hate books with friends
She watch movies on weekends
We play the beach when it rains
They eat movies when it rains
You watch the beach with friends
She play the beach before dinner
My friend play cricket after school
The cat eat cricket every day
I hate apples after school
The dog visit the market in the morning
The cat play the market with friends
The dog prefer the market after school
He eat the market in the summer
I visit the beach during holidays
He love music during holidays
She like books with friends
They play the beach after school
He read movies before dinner
I prefer cricket when it rains
My friend watch movies on weekends
She hate bananas in the morning
You visit music in the morning
My friend hate the park when it rains
My friend play movies with friends
Our teacher enjoy the market during holidays
The dog read football before dinner
My friend play music in the morning
You love football in the summer
They prefer football in the morning
We read the market in the summer
You enjoy the beach every day
The dog enjoy the park in the morning
The cat love movies in the summer
The cat prefer bananas on weekends
Our teacher like cricket during holidays
My friend prefer the park with friends
The dog love apples with friends
They prefer the market after school
They enjoy the beach before dinner
I visit apples after school
We read the market during holidays
The dog prefer the beach before dinner
You like the beach in the morning
She visit the park before dinner
They watch cricket with friends
Our teacher play bananas when it rains
I like music in the summer
They play bananas on weekends
You hate the beach before dinner
Our teacher enjoy football when it rains
She visit the market with friends
We visit books every day
He prefer the park every day
He prefer apples at night
The dog read movies before dinner
She watch books every day
Our teacher read bananas in the morning
My friend eat the beach every day
The dog visit the beach during holidays
They hate apples on weekends
The dog watch football in the morning
They love football during holidays
They hate the market in the summer
My friend love the market on weekends
We eat movies every day
They visit movies in the morning
They prefer the park after school
The cat play the beach before dinner
The cat prefer apples with friends
The dog visit bananas every day
Our teacher prefer movies before dinner
She enjoy the park during holidays
The cat prefer books at night
My friend prefer books with friends
You prefer books before dinner
"""

In [4]:
#TOKENISATION
nltk.download('punkt')
nltk.download('punkt_tab') #punkt is a tokeniser which splits sentences into words based on punctuation abb...

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.


True

In [5]:
tokens = word_tokenize(document.lower())

In [6]:
#buildVocab
vocab = {'<unk>':0} #initialises vocab list int form of 'token' : index number/key

Counter(tokens) #counts the number of times each token appears

Counter({'natural': 1,
         'language': 2,
         'processing': 1,
         '(': 2,
         'nlp': 2,
         ')': 2,
         'is': 1,
         'a': 1,
         'subfield': 1,
         'of': 1,
         'artificial': 1,
         'intelligence': 1,
         'ai': 1,
         '.': 6,
         'it': 119,
         'focuses': 1,
         'on': 115,
         'enabling': 1,
         'computers': 1,
         'to': 1,
         'understand': 1,
         ',': 6,
         'interpret': 1,
         'and': 3,
         'generate': 1,
         'human': 1,
         'combines': 1,
         'computer': 1,
         'science': 1,
         'linguistics': 1,
         'machine': 2,
         'learning': 1,
         "'s": 1,
         'used': 1,
         'in': 198,
         'various': 1,
         'applications': 1,
         'like': 97,
         'translation': 1,
         'sentiment': 1,
         'analysis': 1,
         'chatbots': 1,
         'my': 109,
         'friend': 109,
         'books': 98,
     

In [7]:
for tokens in Counter(tokens).keys():  # .keys is function which only returns unique values from all the tokens and doesnot return duplicate ones
  if tokens not in vocab:
    vocab[tokens] = len(vocab)  #len of vocab calculates the number of tokens in the vocab and assgns the new token(vocab[tokens], the next index)
vocab



{'<unk>': 0,
 'natural': 1,
 'language': 2,
 'processing': 3,
 '(': 4,
 'nlp': 5,
 ')': 6,
 'is': 7,
 'a': 8,
 'subfield': 9,
 'of': 10,
 'artificial': 11,
 'intelligence': 12,
 'ai': 13,
 '.': 14,
 'it': 15,
 'focuses': 16,
 'on': 17,
 'enabling': 18,
 'computers': 19,
 'to': 20,
 'understand': 21,
 ',': 22,
 'interpret': 23,
 'and': 24,
 'generate': 25,
 'human': 26,
 'combines': 27,
 'computer': 28,
 'science': 29,
 'linguistics': 30,
 'machine': 31,
 'learning': 32,
 "'s": 33,
 'used': 34,
 'in': 35,
 'various': 36,
 'applications': 37,
 'like': 38,
 'translation': 39,
 'sentiment': 40,
 'analysis': 41,
 'chatbots': 42,
 'my': 43,
 'friend': 44,
 'books': 45,
 'every': 46,
 'day': 47,
 'our': 48,
 'teacher': 49,
 'read': 50,
 'cricket': 51,
 'during': 52,
 'holidays': 53,
 'i': 54,
 'play': 55,
 'after': 56,
 'school': 57,
 'visit': 58,
 'movies': 59,
 'the': 60,
 'cat': 61,
 'beach': 62,
 'at': 63,
 'night': 64,
 'we': 65,
 'bananas': 66,
 'watch': 67,
 'football': 68,
 'when': 69

In [8]:
len(vocab)

94

In [9]:
#extract sentences
input_sentences = document.split('\n') # breaks and extracts sentences from paras

In [12]:
def text_to_embeddings(sentence, vocab):

  sentence_embedding = []

  for token in sentence:
    if token in vocab:
      sentence_embedding.append(vocab[token])
    else:
      sentence_embedding.append(vocab['<unk>'])
  return sentence_embedding
  #if words of the sentence present in vocab the word's index is added to sentence_embedding
  #if word not in vocab then unk token is sent whose value is 0 so 0 is added to the embedding


In [13]:
text_to_embeddings("i like football", vocab)

[54, 0, 0, 54, 0, 0, 0, 0, 0, 0, 0, 0, 8, 0, 0]

In [14]:
input_numerical_sentences = []
for sentence in input_sentences:
 input_numerical_sentences.append(text_to_embeddings( word_tokenize(sentence.lower()), vocab))

print(input_numerical_sentences)
len(input_numerical_sentences)


[[], [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 4, 13, 6, 14], [15, 16, 17, 18, 19, 20, 21, 22, 23, 22, 24, 25, 26, 2, 14], [5, 27, 28, 29, 22, 30, 22, 24, 31, 32, 14], [15, 33, 34, 35, 36, 37, 38, 31, 39, 22, 40, 41, 22, 24, 42, 14], [43, 44, 38, 45, 46, 47], [48, 49, 50, 51, 52, 53], [54, 55, 51, 56, 57], [43, 44, 58, 59, 56, 57], [60, 61, 55, 60, 62, 63, 64], [65, 58, 66, 63, 64], [65, 67, 68, 69, 15, 70], [60, 61, 50, 60, 62, 71, 72], [73, 50, 51, 46, 47], [43, 44, 74, 75, 76, 77], [78, 79, 45, 76, 77], [43, 44, 79, 59, 76, 77], [54, 79, 75, 46, 47], [60, 61, 80, 68, 63, 64], [60, 81, 74, 66, 17, 82], [73, 58, 60, 62, 17, 82], [60, 81, 38, 45, 71, 72], [65, 67, 75, 71, 72], [60, 61, 38, 45, 76, 77], [65, 74, 60, 83, 46, 47], [43, 44, 67, 60, 62, 52, 53], [48, 49, 84, 51, 52, 53], [85, 67, 66, 56, 57], [54, 80, 75, 17, 82], [43, 44, 50, 59, 35, 60, 86], [48, 49, 87, 75, 46, 47], [60, 81, 55, 60, 88, 46, 47], [60, 61, 67, 45, 46, 47], [89, 84, 60, 83, 35, 60, 90], [60, 61, 74, 45, 35, 6

1004

In [13]:
training_sequence =[]
for sentence in input_numerical_sentences:

  for i in range(1,len(sentence)):
    training_sequence.append(sentence[:i+1])

    #prints sentence by predicting next word one by one
    #starts with 10 = i, 45=am, 21 = rajiv
    #10,45
    #10,45,21

In [14]:
training_sequence

[[1, 2],
 [1, 2, 3],
 [1, 2, 3, 4],
 [1, 2, 3, 4, 5],
 [1, 2, 3, 4, 5, 6],
 [1, 2, 3, 4, 5, 6, 7],
 [1, 2, 3, 4, 5, 6, 7, 8],
 [1, 2, 3, 4, 5, 6, 7, 8, 9],
 [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
 [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11],
 [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12],
 [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 4],
 [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 4, 13],
 [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 4, 13, 6],
 [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 4, 13, 6, 14],
 [15, 16],
 [15, 16, 17],
 [15, 16, 17, 18],
 [15, 16, 17, 18, 19],
 [15, 16, 17, 18, 19, 20],
 [15, 16, 17, 18, 19, 20, 21],
 [15, 16, 17, 18, 19, 20, 21, 22],
 [15, 16, 17, 18, 19, 20, 21, 22, 23],
 [15, 16, 17, 18, 19, 20, 21, 22, 23, 22],
 [15, 16, 17, 18, 19, 20, 21, 22, 23, 22, 24],
 [15, 16, 17, 18, 19, 20, 21, 22, 23, 22, 24, 25],
 [15, 16, 17, 18, 19, 20, 21, 22, 23, 22, 24, 25, 26],
 [15, 16, 17, 18, 19, 20, 21, 22, 23, 22, 24, 25, 26, 2],
 [15, 16, 17, 18, 19, 20, 21, 22, 23, 22, 24, 25, 26, 2, 14],
 [5, 27],
 [

In [15]:
training_sequence[:10]

[[1, 2],
 [1, 2, 3],
 [1, 2, 3, 4],
 [1, 2, 3, 4, 5],
 [1, 2, 3, 4, 5, 6],
 [1, 2, 3, 4, 5, 6, 7],
 [1, 2, 3, 4, 5, 6, 7, 8],
 [1, 2, 3, 4, 5, 6, 7, 8, 9],
 [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
 [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]]

As we can see each trainig sequence is of different size.
we send sentences to the lstm in form of batches usually one batch is of 32 sentences . now if all sentences are of different sizes the training becomes improper so we apply padding in front as the last token becomes the target label for predicting the next word.


In [16]:
len_list = []

for sequence in training_sequence:
  len_list.append(len(sequence))

max(len_list)

16

In [17]:
#so we have to make each sentence of 16 size

for sequence in training_sequence:
  while len(sequence)<max(len_list):
    sequence.insert(0,0)

print(training_sequence)

[[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4, 5], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4, 5, 6], [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4, 5, 6, 7], [0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4, 5, 6, 7, 8], [0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9], [0, 0, 0, 0, 0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10], [0, 0, 0, 0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11], [0, 0, 0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12], [0, 0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 4], [0, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 4, 13], [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 4, 13, 6], [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 4, 13, 6, 14], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 15, 16], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 15, 16, 17], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 15, 16, 17, 18], [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 15, 16, 17, 18, 19], [0, 0, 0, 0, 

In [18]:
len(training_sequence[34])

16

In [19]:
training_sequence = torch.tensor(training_sequence, dtype=torch.long)
#converts the training sequence which is list of lists to a n dimensional matrix which is called a tensor
#tensor is usedful for direct computations as well as autodifferentiation
#dtype long to typecast all values to long

In [20]:
training_sequence

tensor([[ 0,  0,  0,  ...,  0,  1,  2],
        [ 0,  0,  0,  ...,  1,  2,  3],
        [ 0,  0,  0,  ...,  2,  3,  4],
        ...,
        [ 0,  0,  0,  ..., 85, 87, 45],
        [ 0,  0,  0,  ..., 87, 45, 76],
        [ 0,  0,  0,  ..., 45, 76, 77]])

In [21]:
training_sequence.shape

torch.Size([5092, 16])

In [22]:
#[:,:] -> 2D matrix indication containing all rows and all cols
#[:,1] -> all rows 1st col, [:,:-1]-> all rows , all cols except last
X = training_sequence[:,:-1]  #X is the input that is all except the last word
y = training_sequence[:,-1]   #y is the output or the next word which is to be predicted

In [23]:
X


tensor([[ 0,  0,  0,  ...,  0,  0,  1],
        [ 0,  0,  0,  ...,  0,  1,  2],
        [ 0,  0,  0,  ...,  1,  2,  3],
        ...,
        [ 0,  0,  0,  ...,  0, 85, 87],
        [ 0,  0,  0,  ..., 85, 87, 45],
        [ 0,  0,  0,  ..., 87, 45, 76]])

In [24]:
y

tensor([ 2,  3,  4,  ..., 45, 76, 77])

In [25]:
class CustomDataset(Dataset):

  def __init__(self, X, y):
    self.X = X
    self.y = y

  def __len__(self):
    return self.X.shape[0]

  def __getitem__(self, index):
    return self.X[index], self.y[index]

In [26]:
dataset=CustomDataset(X,y)

In [27]:
len(dataset)

5092

In [28]:
dataset[4]

(tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 2, 3, 4, 5]), tensor(6))

In [29]:
dataloader = DataLoader(dataset, batch_size=32, shuffle=True)


In [30]:
for x,y in dataloader:
  print(x,y)
  break

tensor([[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 60, 61, 67],
        [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 60, 81, 79, 68],
        [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 60, 61],
        [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 60, 81, 84],
        [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 89, 50, 68],
        [ 0,  0,  0,  0,  0,  0,  0,  0,  0, 48, 49, 67, 60, 62, 35],
        [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 89, 84],
        [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 73],
        [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 89, 50, 45],
        [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 85, 50],
        [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 48],
        [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 89, 87, 68, 35],
        [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 65, 38, 60, 88],
        [ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 60],
        [ 0,  0,  0,

X is in the shape of (5092,15)
and goes batch wise  (32,15)
        
y is in the shape of (5092)

LSTM Architecture using RNN
the first word of each of the 32 sentences goes into lstm and it returns an output and hidden state which is input to another layer this continues for 15 layers and last output generated is the predicted word. the summations of all weights is taken to calculate loss and add bias to minimize loss



In [31]:
from typing_extensions import final
class LSTMModel(nn.Module):
  def __init__(self, vocab_size):
    super().__init__()
    self.embedding = nn.Embedding(vocab_size, 100)
    self.lstm = nn.LSTM(100, 128, batch_first=True)
    self.fc = nn.Linear(128, vocab_size)

  def forward(self, x):
    embedded = self.embedding(x)
    intermediate_hidden_states, (final_hidden_state, final_cell_state) = self.lstm(embedded)
    output = self.fc(final_hidden_state.squeeze(0))
    return output

In [37]:
model = LSTMModel(len(vocab))

In [38]:
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

In [39]:
model.to(device)

LSTMModel(
  (embedding): Embedding(94, 100)
  (lstm): LSTM(100, 128, batch_first=True)
  (fc): Linear(in_features=128, out_features=94, bias=True)
)

In [40]:
epochs = 50
learning_rate = 0.001

criterion = nn.CrossEntropyLoss()

optimiser = torch.optim.Adam(model.parameters(), lr=learning_rate)

In [42]:
for epoch in range(epochs):
  total_loss = 0

  for batch_x,batch_y in dataloader:
    batch_x = batch_x.to(device)
    batch_y = batch_y.to(device)

    optimiser.zero_grad()

    output = model(batch_x)

    loss = criterion(output,batch_y)

    loss.backward()

    optimiser.step()

    total_loss += loss.item()

  print(f"Epoch: {epoch+1}, Loss:{total_loss:.4f}")


Epoch: 1, Loss:168.7852
Epoch: 2, Loss:168.5059
Epoch: 3, Loss:166.8867
Epoch: 4, Loss:166.6770
Epoch: 5, Loss:165.6895
Epoch: 6, Loss:166.2468
Epoch: 7, Loss:164.8091
Epoch: 8, Loss:164.9783
Epoch: 9, Loss:164.4402
Epoch: 10, Loss:163.5234
Epoch: 11, Loss:164.7236
Epoch: 12, Loss:163.3269
Epoch: 13, Loss:162.8353
Epoch: 14, Loss:162.6776
Epoch: 15, Loss:162.7427
Epoch: 16, Loss:161.7902
Epoch: 17, Loss:161.5204
Epoch: 18, Loss:161.9242
Epoch: 19, Loss:160.8630
Epoch: 20, Loss:160.7052
Epoch: 21, Loss:161.5417
Epoch: 22, Loss:160.9336
Epoch: 23, Loss:160.8632
Epoch: 24, Loss:159.9935
Epoch: 25, Loss:160.7474
Epoch: 26, Loss:159.7361
Epoch: 27, Loss:160.7177
Epoch: 28, Loss:160.3058
Epoch: 29, Loss:159.5374
Epoch: 30, Loss:160.3953
Epoch: 31, Loss:159.5208
Epoch: 32, Loss:159.1349
Epoch: 33, Loss:159.0339
Epoch: 34, Loss:158.5723
Epoch: 35, Loss:159.4329
Epoch: 36, Loss:159.2234
Epoch: 37, Loss:158.6116
Epoch: 38, Loss:157.9670
Epoch: 39, Loss:159.2256
Epoch: 40, Loss:158.4953
Epoch: 41

In [52]:
def prediction(model, vocab, text):
  #text to numerical values
  tokenised_text = word_tokenize(text.lower())
  print(tokenised_text)

  numerical_text = text_to_embeddings(tokenised_text, vocab)
  print(numerical_text)

  #padding
  padded_text = torch.tensor([0]*(15 - len(numerical_text)) + numerical_text, dtype = torch.long).unsqueeze(0)
  print(padded_text)

  #send to model-> o/p
  output  = model(padded_text)

  #predicted index
  value,index = torch.max(output, dim = 1)
  print(index)

  #merge with text
  return text+' '+list(vocab.keys())[index]



In [51]:
prediction(model,vocab,"I want")

['i', 'want']
[54, 0]
tensor([[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 54,  0]])
tensor([55])


'I want play'

In [53]:
num_tokens = 10
input_text = "Our teacher"

for i in range(num_tokens):
  predicted_text = prediction(model, vocab, input_text)
  input_text = predicted_text
  print(input_text)


['our', 'teacher']
[48, 49]
tensor([[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 48, 49]])
tensor([67])
Our teacher watch
['our', 'teacher', 'watch']
[48, 49, 67]
tensor([[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 48, 49, 67]])
tensor([60])
Our teacher watch the
['our', 'teacher', 'watch', 'the']
[48, 49, 67, 60]
tensor([[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 48, 49, 67, 60]])
tensor([88])
Our teacher watch the market
['our', 'teacher', 'watch', 'the', 'market']
[48, 49, 67, 60, 88]
tensor([[ 0,  0,  0,  0,  0,  0,  0,  0,  0,  0, 48, 49, 67, 60, 88]])
tensor([46])
Our teacher watch the market every
['our', 'teacher', 'watch', 'the', 'market', 'every']
[48, 49, 67, 60, 88, 46]
tensor([[ 0,  0,  0,  0,  0,  0,  0,  0,  0, 48, 49, 67, 60, 88, 46]])
tensor([47])
Our teacher watch the market every day
['our', 'teacher', 'watch', 'the', 'market', 'every', 'day']
[48, 49, 67, 60, 88, 46, 47]
tensor([[ 0,  0,  0,  0,  0,  0,  0,  0, 48, 49, 67, 60, 88, 46, 47]])
tensor([47